api/catalog/api/controllers/elasticsearch/search.py

```diff
@@ -0,0 +1,198 @@
+from elasticsearch_dsl import Q, Search
+from elasticsearch_dsl.response import Hit
+
+from catalog.api.controllers.elasticsearch.utils import (
+    exclude_filtered_providers,
+    get_query_slice,
+    get_result_and_page_count,
+    post_process_results,
+)
+from catalog.api.serializers.media_serializers import MediaSearchRequestSerializer
+
+
+def _quote_escape(query_string: str) -> str:
+    """
+    If there are any unmatched quotes in the query supplied by the user, ignore
+    them by escaping.
+
+    :param query_string: the string in which to escape unbalanced quotes
+    :return: the given string, if the quotes are balanced, the escaped string otherwise
+    """
+
+    num_quotes = query_string.count('"')
+    if num_quotes % 2 == 1:
+        return query_string.replace('"', '\\"')
+    else:
+        return query_string
+
+
+def _apply_filter(
+    s: Search,
+    query_serializer: MediaSearchRequestSerializer,
+    basis: Union[str, tuple[str, str]],
+    behaviour: Literal["filter", "exclude"] = "filter",
+) -> Search:
+    """
+    Parse and apply a filter from the search parameters serializer. The
+    parameter key is assumed to have the same name as the corresponding
+    Elasticsearch property. Each parameter value is assumed to be a comma
+    separated list encoded as a string.
+
+    :param s: the search query to issue to Elasticsearch
+    :param query_serializer: the ``MediaSearchRequestSerializer`` object with the query
+    :param basis: the name of the field in the serializer and Elasticsearch
+    :param behaviour: whether to accept (``filter``) or reject (``exclude``) the hit
+    :return: the modified search query
+    """
+
+    search_params = query_serializer.data
+    if isinstance(basis, tuple):
+        serializer_field, es_field = basis
+    else:
+        serializer_field = es_field = basis
+    if serializer_field in search_params:
+        filters = []
+        for arg in search_params[serializer_field].split(","):
+            filters.append(Q("term", **{es_field: arg}))
+        method = getattr(s, behaviour)  # can be ``s.filter`` or ``s.exclude``
+        return method("bool", should=filters)
+    else:
+        return s
+
+
+def perform_search(
+    query_serializer: MediaSearchRequestSerializer,
+    index: Literal["image", "audio"],
+    ip: int,
+) -> Tuple[List[Hit], int, int]:
+    """
+    Perform a ranked, paginated search based on the query and filters given in the
+    search request.
+
+    :param query_serializer: the ``MediaSearchRequestSerializer`` object with the query
+    :param index: The Elasticsearch index to search (e.g. 'image')
+    :param ip: the users' hashed IP to consistently route to the same ES shard
+    :return: the list of search results with the page and result count
+    """
+
+    s = Search(using="default", index=index)
+    search_params = query_serializer.data
+
+    rules: dict[Literal["filter", "exclude"], list[Union[str, tuple[str, str]]]] = {
```

Comment:
nit: With such a specific structure it seems like maybe a group of dataclasses configured in this fashion could be (a) more robust in the face of future refactors and (b) clearer for someone adding a new aspect to this configuration. It'd also give each thing a clearer and individual name (and a docstring explaining why the name):

```py
from dataclasses import dataclass, field


@dataclass
class Basis:
    """https://stackoverflow.com/a/61480946"""
    name: str
    _field: Optional[str] = field(init=False, repr=False, default=None)
    
    @property
    def field(self) -> str:
	return self._field or self.name
    
    @field.setter
    def field(self, value: Optional[str] = None) -> None:
	self._field = value
	

@dataclass
class Rules:
    filter: list[Basis]
    exclude: list[Basis]
    
    def get(self) -> Mapping[Literal['filter', 'exclude'], list[Basis]]:
	return {
	    "filter": self.filter,
	    "exclude": self.exclude
	}


rules = Rules(
    filter=[
	Basis("title"),
	Basis("license", "license.keyword"),
	# ... etc
    ],
    exclude=[
	Basis("excluded_source", "source"),
    ]
)


for behaviour, bases in rules.get():
    for basis in bases:
	s = _apply_filter(s, query_serializer, basis, behaviour)
```

It would give a natural (and central/easy to find) place to explain or cite the origin of the name "basis" as well.

Reply:
Using a data-class for `FieldMapping` (`Basis` in your example) was a bit too verbose so I used a regular class for that purpose. Hopefully the new code in 0c86c03 is much easier to grasp.